Oracle StoredProcedure call with cursor in native mode raises a NullPointerException

Calling an Oracle StoredProcedure with cursor as out parameter raises a NullPointerException in native mode:
```
Caused by: java.lang.NullPointerException
       at org.hibernate.engine.jdbc.cursor.internal.StandardRefCursorSupport.registerRefCursorParameter(StandardRefCursorSupport.java:67)
       at org.hibernate.query.procedure.internal.ProcedureParameterImpl.prepare(ProcedureParameterImpl.java:283)
       at org.hibernate.procedure.internal.ProcedureCallImpl$3.accept(ProcedureCallImpl.java:397)
```
The reproducer (a maven project) is available at https://github.com/draccagni/quarkus-hibernate-orm-oracle.git . All the instructions are explained in the ```README.md``` file

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/ClassNames.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/ClassNames.java
@@ -42,6 +42,7 @@ public class ClassNames {
     public static final DotName TYPE = createConstant("org.hibernate.annotations.Type");
     public static final DotName TYPE_DEFINITION = createConstant("org.hibernate.annotations.TypeDef");
     public static final DotName TYPE_DEFINITIONS = createConstant("org.hibernate.annotations.TypeDefs");
+    public static final DotName INJECT_SERVICE = createConstant("org.hibernate.service.spi.InjectService");
 
     public static final DotName ENTITY_MANAGER_FACTORY = createConstant("javax.persistence.EntityManagerFactory");
     public static final DotName SESSION_FACTORY = createConstant("org.hibernate.SessionFactory");

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmAnnotations.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmAnnotations.java
@@ -216,4 +216,9 @@ public final class HibernateOrmAnnotations {
             DotName.createSimple("org.hibernate.annotations.ValueGenerationType"),
             DotName.createSimple("org.hibernate.annotations.Where"),
             DotName.createSimple("org.hibernate.annotations.WhereJoinTable")));
+
+    public static final List<DotName> ANNOTATED_WITH_INJECT_SERVICE = Collections.unmodifiableList(Arrays.asList(
+            DotName.createSimple("org.hibernate.engine.jdbc.connections.internal.DatasourceConnectionProviderImpl"),
+            DotName.createSimple("org.hibernate.engine.jdbc.cursor.internal.StandardRefCursorSupport")));
+
 }

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
@@ -664,6 +664,30 @@ public final class HibernateOrmProcessor {
         }
     }
 
+    /*
+     * Enable reflection for methods annotated with @InjectService,
+     * such as org.hibernate.engine.jdbc.cursor.internal.StandardRefCursorSupport.injectJdbcServices.
+     */
+    @BuildStep(onlyIf = NativeOrNativeSourcesBuild.class)
+    public void registerInjectServiceMethodsForReflection(CombinedIndexBuildItem index,
+            BuildProducer<ReflectiveClassBuildItem> reflective) {
+        Set<String> classes = new HashSet<>();
+
+        // Built-in service classes; can't rely on Jandex as Hibernate ORM is not indexed by default.
+        HibernateOrmAnnotations.ANNOTATED_WITH_INJECT_SERVICE.stream()
+                .map(DotName::toString)
+                .forEach(classes::add);
+
+        // Integrators relying on @InjectService.
+        index.getIndex().getAnnotations(ClassNames.INJECT_SERVICE).stream()
+                .map(a -> a.target().asMethod().declaringClass().name().toString())
+                .forEach(classes::add);
+
+        if (!classes.isEmpty()) {
+            reflective.produce(new ReflectiveClassBuildItem(false, true, false, classes.toArray(new String[0])));
+        }
+    }
+
     private static Optional<String> getSqlLoadScript(Optional<String> sqlLoadScript, LaunchMode launchMode) {
         // Explicit file or default Hibernate ORM file.
         if (sqlLoadScript.isPresent()) {

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/HibernateOrmAnnotationsTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/HibernateOrmAnnotationsTest.java
@@ -26,6 +26,7 @@ import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 
 import io.quarkus.deployment.index.IndexingUtil;
+import io.quarkus.hibernate.orm.deployment.ClassNames;
 import io.quarkus.hibernate.orm.deployment.HibernateOrmAnnotations;
 
 /**
@@ -74,6 +75,15 @@ public class HibernateOrmAnnotationsTest {
                 .containsExactlyInAnyOrderElementsOf(packageLevelHibernateAnnotations);
     }
 
+    @Test
+    public void testNoMissingInjectServiceAnnotatedClass() {
+        Set<DotName> injectServiceAnnotatedClasses = findClassesWithMethodsAnnotatedWith(hibernateIndex,
+                ClassNames.INJECT_SERVICE);
+
+        assertThat(HibernateOrmAnnotations.ANNOTATED_WITH_INJECT_SERVICE)
+                .containsExactlyInAnyOrderElementsOf(injectServiceAnnotatedClasses);
+    }
+
     private Set<DotName> findRuntimeAnnotations(Index index) {
         Set<DotName> annotations = new HashSet<>();
         for (AnnotationInstance retentionAnnotation : index.getAnnotations(RETENTION)) {
@@ -95,6 +105,15 @@ public class HibernateOrmAnnotationsTest {
             }
         }
         return annotations;
+    }
+
+    private Set<DotName> findClassesWithMethodsAnnotatedWith(Index index, DotName annotationName) {
+        Set<DotName> classes = new TreeSet<>();
+        for (AnnotationInstance annotation : index.getAnnotations(annotationName)) {
+            ClassInfo clazz = annotation.target().asMethod().declaringClass();
+            classes.add(clazz.name());
+        }
+        return classes;
     }
 
     private boolean allowsTargetType(ClassInfo annotation, ElementType targetType) {

--- a/integration-tests/jpa-oracle/pom.xml
+++ b/integration-tests/jpa-oracle/pom.xml
@@ -47,6 +47,11 @@
             <artifactId>rest-assured</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
 
         <!-- Minimal test dependencies to *-deployment artifacts for consistent build order -->
         <dependency>

--- a/integration-tests/jpa-oracle/src/main/java/io/quarkus/example/jpaoracle/procedurecall/ProcedureCallEndpoint.java
+++ b/integration-tests/jpa-oracle/src/main/java/io/quarkus/example/jpaoracle/procedurecall/ProcedureCallEndpoint.java
@@ -1,0 +1,82 @@
+package io.quarkus.example.jpaoracle.procedurecall;
+
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+import javax.persistence.ParameterMode;
+import javax.persistence.StoredProcedureQuery;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.transaction.Transactional;
+
+import io.agroal.api.AgroalDataSource;
+import io.quarkus.arc.Arc;
+import io.quarkus.runtime.StartupEvent;
+
+@WebServlet("/jpa-oracle/procedure-call/")
+public class ProcedureCallEndpoint extends HttpServlet {
+
+    private static final String PROCEDURE_NAME = "myproc";
+
+    @Inject
+    EntityManager em;
+
+    void installStoredProcedure(@Observes StartupEvent ev) throws SQLException {
+        AgroalDataSource dataSource = Arc.container().instance(AgroalDataSource.class).get();
+        try (Connection conn = dataSource.getConnection()) {
+            try (Statement st = conn.createStatement()) {
+                st.execute("CREATE OR REPLACE PROCEDURE " + PROCEDURE_NAME
+                        + " (p_pattern IN VARCHAR2, p_cur OUT SYS_REFCURSOR)\n"
+                        + "AS\n"
+                        + "BEGIN\n"
+                        + "   OPEN p_cur FOR SELECT name FROM " + ProcedureCallEntity.NAME + " WHERE name LIKE p_pattern;\n"
+                        + "END " + PROCEDURE_NAME + ";");
+            }
+        }
+    }
+
+    @Transactional
+    void initData(@Observes StartupEvent ev) {
+        persistEntity("prefix#1");
+        persistEntity("other#1");
+        persistEntity("prefix#2");
+        persistEntity("other#2");
+    }
+
+    private void persistEntity(String name) {
+        ProcedureCallEntity entity = new ProcedureCallEntity();
+        entity.setName(name);
+        em.persist(entity);
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        try {
+            String pattern = req.getParameter("pattern");
+            StoredProcedureQuery storedProcedure = em.createStoredProcedureQuery(PROCEDURE_NAME);
+            storedProcedure.registerStoredProcedureParameter("p_pattern", String.class, ParameterMode.IN);
+            storedProcedure.registerStoredProcedureParameter("p_cur", Object.class, ParameterMode.REF_CURSOR);
+            storedProcedure.setParameter("p_pattern", pattern);
+            storedProcedure.execute();
+
+            List<String> result = new ArrayList<>();
+            ResultSet resultSet = (ResultSet) storedProcedure.getOutputParameterValue("p_cur");
+            while (resultSet.next()) {
+                result.add(resultSet.getString(1));
+            }
+            resp.getWriter().write(String.join("\n", result));
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+}

--- a/integration-tests/jpa-oracle/src/main/java/io/quarkus/example/jpaoracle/procedurecall/ProcedureCallEntity.java
+++ b/integration-tests/jpa-oracle/src/main/java/io/quarkus/example/jpaoracle/procedurecall/ProcedureCallEntity.java
@@ -1,0 +1,40 @@
+package io.quarkus.example.jpaoracle.procedurecall;
+
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+
+@Entity(name = ProcedureCallEntity.NAME)
+public class ProcedureCallEntity {
+
+    public static final String NAME = "procentity";
+
+    @Id
+    @GeneratedValue
+    private Long id;
+
+    private String name;
+
+    public ProcedureCallEntity() {
+    }
+
+    public ProcedureCallEntity(String name) {
+        this.name = name;
+    }
+
+    public Long getId() {
+        return id;
+    }
+
+    public void setId(Long id) {
+        this.id = id;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+}

--- a/integration-tests/jpa-oracle/src/test/java/io/quarkus/it/jpa/oracle/procedurecall/ProcedureCallInGraalITCase.java
+++ b/integration-tests/jpa-oracle/src/test/java/io/quarkus/it/jpa/oracle/procedurecall/ProcedureCallInGraalITCase.java
@@ -1,0 +1,7 @@
+package io.quarkus.it.jpa.oracle.procedurecall;
+
+import io.quarkus.test.junit.NativeImageTest;
+
+@NativeImageTest
+public class ProcedureCallInGraalITCase extends ProcedureCallTest {
+}

--- a/integration-tests/jpa-oracle/src/test/java/io/quarkus/it/jpa/oracle/procedurecall/ProcedureCallTest.java
+++ b/integration-tests/jpa-oracle/src/test/java/io/quarkus/it/jpa/oracle/procedurecall/ProcedureCallTest.java
@@ -1,0 +1,30 @@
+package io.quarkus.it.jpa.oracle.procedurecall;
+
+import static io.restassured.RestAssured.given;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+
+/**
+ * Test that procedure calls work correctly, in particular when cursors are involved.
+ * <p>
+ * They used to fail in native mode because adding cursor parameters involved reflection during Hibernate ORM startup,
+ * and the corresponding methods were not registered for reflection.
+ * See https://github.com/quarkusio/quarkus/issues/17295.
+ */
+@QuarkusTest
+public class ProcedureCallTest {
+
+    @Test
+    public void test() {
+        String response = given().param("pattern", "prefix%")
+                .when().get("/jpa-oracle/procedure-call/").then()
+                .statusCode(200)
+                .extract().body().asString();
+        assertThat(response.split("\n"))
+                .containsExactlyInAnyOrder("prefix#1", "prefix#2");
+    }
+
+}

